### ISSUE TITLE
Current ES version (0.90.2) is not compatible with current Titan version (0.3.2)

### ISSUE THREAD

**Issue by dkuppitz** [thinkaurelius/titan#339]
The following script works as expected in Titan 0.3.2 / ES 0.90.0, but crashes with Titan 0.3.2 / ES 0.90.2.

``` groovy
config = new BaseConfiguration()
config.setProperty("storage.backend", "inmemory")
config.setProperty("storage.index.search.backend", "elasticsearch")
config.setProperty("storage.index.search.hostname", "10.0.0.30")
config.setProperty("storage.index.search.client-only", true)

g = TitanFactory.open(config)
g.makeType().name("geo").dataType(Geoshape.class).unique(OUT).indexed("search", Vertex.class).makePropertyKey()
g.commit()

home = g.addVertex(["place":"home"])
g.commit()

home.setProperty("geo", Geoshape.point(51.012217,6.077343))
g.commit()
```

**Stacktrace for ES 0.90.2**

```
gremlin> g.commit()
13/07/30 21:00:38 ERROR consistentkey.LocalLockMediator: Local unlock failed: no locks found for KeyColumn [k=0x0-0-0-0-0-0-0-8, c=0x1-138]
13/07/30 21:00:38 WARN consistentkey.ConsistentKeyLockTransaction: Failed to release LockClaim [backer=com.thinkaurelius.titan.diskstorage.locking.consistentkey.ConsistentKeyLockStore@73974028, key=0x0-0-0-0-0-0-0-8, col=0x1-138, expectedValue=null] locally (txn=com.thinkaurelius.titan.diskstorage.locking.consistentkey.ConsistentKeyLockTransaction@236c6da9)
Could not commit transaction due to exception during persistence
Display stack trace? [yN] y
com.thinkaurelius.titan.core.TitanException: Could not commit transaction due to exception during persistence
        at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:848)
        at com.thinkaurelius.titan.graphdb.blueprints.TitanBlueprintsGraph.commit(TitanBlueprintsGraph.java:64)
        at com.tinkerpop.blueprints.TransactionalGraph$commit.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:42)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:112)
        at groovysh_evaluate.run(groovysh_evaluate:56)
        at groovysh_evaluate$run.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:42)
        at groovysh_evaluate$run.call(Unknown Source)
        at org.codehaus.groovy.tools.shell.Interpreter.evaluate(Interpreter.groovy:67)
        at org.codehaus.groovy.tools.shell.Interpreter$evaluate.call(Unknown Source)
        at org.codehaus.groovy.tools.shell.Groovysh.execute(Groovysh.groovy:152)
        at org.codehaus.groovy.tools.shell.Shell.leftShift(Shell.groovy:114)
        at org.codehaus.groovy.tools.shell.Shell$leftShift$0.call(Unknown Source)
        at org.codehaus.groovy.tools.shell.ShellRunner.work(ShellRunner.groovy:88)
        at org.codehaus.groovy.tools.shell.InteractiveShellRunner.super$2$work(InteractiveShellRunner.groovy)
        at sun.reflect.GeneratedMethodAccessor37.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:616)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:233)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1079)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodOnSuperN(ScriptBytecodeAdapter.java:128)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodOnSuper0(ScriptBytecodeAdapter.java:148)
        at org.codehaus.groovy.tools.shell.InteractiveShellRunner.work(InteractiveShellRunner.groovy:100)
        at sun.reflect.GeneratedMethodAccessor36.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:616)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:272)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:52)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:137)
        at org.codehaus.groovy.tools.shell.ShellRunner.run(ShellRunner.groovy:57)
        at org.codehaus.groovy.tools.shell.InteractiveShellRunner.super$2$run(InteractiveShellRunner.groovy)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:616)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:233)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1079)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodOnSuperN(ScriptBytecodeAdapter.java:128)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodOnSuper0(ScriptBytecodeAdapter.java:148)
        at org.codehaus.groovy.tools.shell.InteractiveShellRunner.run(InteractiveShellRunner.groovy:66)
        at com.thinkaurelius.titan.tinkerpop.gremlin.Console.<init>(Console.java:65)
        at com.thinkaurelius.titan.tinkerpop.gremlin.Console.<init>(Console.java:78)
        at com.thinkaurelius.titan.tinkerpop.gremlin.Console.main(Console.java:104)
Caused by: com.thinkaurelius.titan.diskstorage.PermanentStorageException: Unknown exception while executing index operation
        at com.thinkaurelius.titan.diskstorage.es.ElasticSearchIndex.convert(ElasticSearchIndex.java:169)
        at com.thinkaurelius.titan.diskstorage.es.ElasticSearchIndex.mutate(ElasticSearchIndex.java:305)
        at com.thinkaurelius.titan.diskstorage.indexing.IndexTransaction.flushInternal(IndexTransaction.java:88)
        at com.thinkaurelius.titan.diskstorage.indexing.IndexTransaction.commit(IndexTransaction.java:70)
        at com.thinkaurelius.titan.diskstorage.BackendTransaction.commit(BackendTransaction.java:71)
        at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:841)
        ... 46 more
Caused by: org.elasticsearch.common.util.concurrent.UncategorizedExecutionException: Failed execution
        at org.elasticsearch.action.support.AdapterActionFuture.rethrowExecutionException(AdapterActionFuture.java:88)
        at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:49)
        at com.thinkaurelius.titan.diskstorage.es.ElasticSearchIndex.mutate(ElasticSearchIndex.java:298)
        ... 50 more
Caused by: java.lang.IndexOutOfBoundsException: Readable byte limit exceeded: 263
        at org.elasticsearch.common.netty.buffer.AbstractChannelBuffer.readByte(AbstractChannelBuffer.java:236)
        at org.elasticsearch.transport.netty.ChannelBufferStreamInput.readByte(ChannelBufferStreamInput.java:132)
        at org.elasticsearch.common.io.stream.AdapterStreamInput.readByte(AdapterStreamInput.java:35)
        at org.elasticsearch.common.io.stream.StreamInput.readBoolean(StreamInput.java:252)
        at org.elasticsearch.action.update.UpdateRequest.readFrom(UpdateRequest.java:566)
        at org.elasticsearch.transport.netty.MessageChannelHandler.handleRequest(MessageChannelHandler.java:204)
        at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:108)
        at org.elasticsearch.common.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
        at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:296)
        at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
        at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
        at org.elasticsearch.common.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
        at org.elasticsearch.common.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
        at org.elasticsearch.common.netty.OpenChannelsHandler.handleUpstream(OpenChannelsHandler.java:74)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
        at org.elasticsearch.common.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
        at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:268)
        at org.elasticsearch.common.netty.channel.Channels.fireMessageReceived(Channels.java:255)
        at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:109)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:312)
        at org.elasticsearch.common.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:90)
        at org.elasticsearch.common.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
        at org.elasticsearch.common.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
        at org.elasticsearch.common.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:724)
```

Interesting fact: The following script works well in both ES versions:

``` groovy
config = new BaseConfiguration()
config.setProperty("storage.backend", "inmemory")
config.setProperty("storage.index.search.backend", "elasticsearch")
config.setProperty("storage.index.search.hostname", "10.0.0.30")
config.setProperty("storage.index.search.client-only", true)

g = TitanFactory.open(config)
g.makeType().name("geo").dataType(Geoshape.class).unique(OUT).indexed("search", Vertex.class).makePropertyKey()
g.commit()

g.addVertex(["place":"home","geo":Geoshape.point(51.012217,6.077343)])
g.commit()
```

So the error only occurs when you update an existing vertex.

**Comment by jcamins** [comment]
I am seeing the same problem with ElasticSearch 0.90.3 on Mac OS X 10.8.4.

**Comment by mbroecheler** [comment]
This exception is so deep inside ElasticSearch that I doubt Titan has anything to do with it. We will keep investigating to see what the issue might be.

**Comment by jcamins** [comment]
Actually, I have a little more information. I just tried removing all the ElasticSearch 0.90.0 JARs from the client and replacing them the JARs from ES 0.90.3. For me this gets it through the first commit, but the second (I commit every thousand records) fails with a new error:
Caused by: java.lang.NoSuchMethodError: org.elasticsearch.action.update.UpdateRequestBuilder.setUpsertRequest(Lorg/elasticsearch/common/xcontent/XContentBuilder;)Lorg/elasticsearch/action/update/UpdateRequestBuilder;
    at com.thinkaurelius.titan.diskstorage.es.ElasticSearchIndex.mutate(ElasticSearchIndex.java:296)
    at com.thinkaurelius.titan.diskstorage.indexing.IndexTransaction.flushInternal(IndexTransaction.java:88)
    at com.thinkaurelius.titan.diskstorage.indexing.IndexTransaction.commit(IndexTransaction.java:70)
    at com.thinkaurelius.titan.diskstorage.BackendTransaction.commit(BackendTransaction.java:71)
    at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:841)
    at com.thinkaurelius.titan.graphdb.blueprints.TitanBlueprintsGraph.commit(TitanBlueprintsGraph.java:64)
    ... 4 more

Does changing the JARs work for you, Daniel?

I looked at org/elasticsearch/common/xcontent/XContentBuilder and it looks like setUpsertRequest may have been renamed to upsert. Changing the method call to upsert() would break compatibility with ES 0.90.0-0.90.2, so is this something you would consider at all?

**Comment by freakinruben** [comment]
Is there any news on becoming compatible with the index api of ElasitcSearcg >= 0.90.3?

**Comment by dalaro** [comment]
I updated master to 0.90.5 in 82ee52069ab84c0359018a4d13090f29b90d17fc.  As noted by @jcamins, it seems like `setUpsertRequest(...)` was renamed `setUpsert(...)`.  That was the only source change I made.

After the upgrade, @dkuppitz's script runs without exceptions on my box.  To get the script to execute, I did change the config to run embedded ES and I replaced unique(...) with vertexUnique(...).  I don't think I modified anything else.
